### ISSUE TITLE
vw core CLI: --nn data-leak: SGD update is incorrect

### ISSUE THREAD

**Issue by arielf** [VowpalWabbit/vowpal_wabbit#4614]
### Describe the bug

Using --nn causes an unexpected data-leak between separate name-spaces
The update goes the wrong way (against the desired gradient towards minimum loss).

### How to reproduce

Full code and data to reproduce is here:

https://github.com/arielf/vw-bugs/tree/main/nn-data-leak

### Version

9.7.0 (git commit: fc9ab2549)

### OS

Linux, Ubuntu 20.04

### Language

CLI/ C++

### Additional context

I tried my best to explain the bug here.

https://github.com/arielf/vw-bugs/tree/main/nn-data-leak

Would appreciate taking a look, thanks!

**Comment by ataymano** [comment]
Can be reproduced on simpler configuration:
data:
```
4.0 always4|always4  f1:1
2.0 always2|always2  f1:1
4.0 always4|always4  f1:1
2.0 always2|always2  f1:1
4.0 always4|always4  f1:1
2.0 always2|always2  f1:1
4.0 always4|always4  f1:1
2.0 always2|always2  f1:1
4.0 always4|always4  f1:1
2.0 always2|always2  f1:1
```

command line:
```
--noconstant --learning_rate 2 --nn 1
```

**Comment by ataymano** [comment]
Can you please elaborate your suggestion?
Updates to weights that are associated to features are already isolated, but there is a problem with last layer constant.
1) It is getting updated on every example and affecting every prediction
2) It is not affected by --noconstant flag

```
from vowpalwabbit import pyvw
import json

vw = pyvw.Workspace('--noconstant --nn 1 --dump_json_weights_include_feature_names_experimental --invert_hash inv.txt')

vw.learn('1 |a x:1')
print('-----------AFTER SEEING a/x-------------------')
print(json.dumps(json.loads(vw.json_weights()), indent=2))

vw.learn('1 |b x:1')
print('-----------AFTER SEEING b/x-------------------')
print(json.dumps(json.loads(vw.json_weights()), indent=2))
```
Output (please see the fact that "a/x", first HiddenBias and first OutputWeight stay the same after seeing second example, but OutputLayerConst has changed):
```
-----------AFTER SEEING a/x-------------------
{
  "weights": [
    {
      "terms": [
        {
          "name": "HiddenBias",
          "namespace": "",
          "string_value": null
        }
      ],
      "offset": 0,
      "index": 29015,
      "value": -0.31610044836997986
    },
    {
      "terms": [
        {
          "name": "x",
          "namespace": "a",
          "string_value": null
        }
      ],
      "offset": 0,
      "index": 110864,
      "value": -0.24449872970581055
    },
    {
      "terms": [
        {
          "name": "OutputWeight",
          "namespace": "",
          "string_value": null
        }
      ],
      "offset": 1,
      "index": 156908,
      "value": -0.3023429214954376
    },
    {
      "terms": [
        {
          "name": "OutputLayerConst",
          "namespace": "",
          "string_value": null
        }
      ],
      "offset": 1,
      "index": 156909,
      "value": 0.3933953046798706
    }
  ]
}
-----------AFTER SEEING b/x-------------------
{
  "weights": [
    {
      "terms": [
        {
          "name": "x",
          "namespace": "b",
          "string_value": null
        }
      ],
      "offset": 0,
      "index": 24482,
      "value": -0.1713941991329193
    },
    {
      "terms": [
        {
          "name": "HiddenBias",
          "namespace": "",
          "string_value": null
        }
      ],
      "offset": 0,
      "index": 29015,
      "value": -0.31610044836997986
    },
    {
      "terms": [
        {
          "name": "x",
          "namespace": "a",
          "string_value": null
        }
      ],
      "offset": 0,
      "index": 110864,
      "value": -0.24449872970581055
    },
    {
      "terms": [
        {
          "name": "OutputWeight",
          "namespace": "",
          "string_value": null
        }
      ],
      "offset": 1,
      "index": 156908,
      "value": -0.3023429214954376
    },
    {
      "terms": [
        {
          "name": "OutputLayerConst",
          "namespace": "",
          "string_value": null
        }
      ],
      "offset": 1,
      "index": 156909,
      "value": 0.6044306755065918
    }
  ]
}
```

**Comment by arielf** [comment]
Hi Alexey,

First, thanks so much for looking into this.

I understand that the hidden layer currently has no notion of name-space isolation.

Not sure if this is too involved (won't fix, or "it is a feature, not a bug") but ideally,
Assuming "anything is possible", there would be a new switch (`--restricted_updates`)  when `--nn` is in effect.

This switch would modify the current behavior to skip (not update) any hidden-layer pairing of features where any of the two features crossed is not in the present example name-space.

Is this clearer?  Too hard/involved to implement?

This is not so critical for me.  I figured out a way around the issue.  Just thought it would be valuable to report this anyway since this behavior is unexpected in the ways demonstrated: (1) non-monotonic convergence (2) behavior unlike with other reductions.

Thanks again!

**Comment by ataymano** [comment]
It sounds like it means that hidden layer is never going to be updated? But maybe I am missing something.
We have [open meeting](https://[vowpalwabbit.org/blog/open-meeting.html](https://vowpalwabbit.org/blog/open-meeting.html)) every other Thursday (including one this week) - do you want to join and maybe discuss this proposal with the team?
Also, as a workaround (maybe?) - nn can passthrough inputs directly to the last layer (--inpass) - it is not guaranteeing separation that you need but still increasing the chance of it to happen (at least simple example like the one in this issue bceome monotonic).

**Comment by arielf** [comment]
Fixed the link, which wasn't working. Thanks for the invite. I can't make it.
So from what you say, I understand each node in the hidden layer is fully connected (to all possible inputs).
If so, then you're correct, and my previous "idea" wouldn't work.  `--inpass` already exists as an sub-option to `--nn`, but it is not disabling passing via the hidden-layer, (it is in addition) correct? Maybe a mode doing only the `--inpass` would be a useful option. I do want crossing of features, but only if they are in the same name-space (like other vw algos work)

I realize this may not be trivial.

**Comment by JohnLangford** [comment]
## Investigation Summary

I've analyzed the neural network reduction code (`nn.cc`) in detail. This is an **architectural limitation** rather than a straightforward bug.

### Root Cause

The `--nn` reduction implements a standard **fully-connected neural network**:

1. **Line 205:** `base.multipredict(ec, 0, n.k, hidden_units, true)` - ALL input features from ALL namespaces connect to ALL hidden units
2. **Lines 347-364:** `base.update(ec, i)` - Weight updates use the full feature gradient, affecting weights for ALL features
3. **Lines 254-318:** Output layer combines ALL hidden unit activations

This is the expected behavior for a fully-connected neural network - the whole point is that all inputs can influence all outputs through learned weights.

### Why Namespace "Leakage" Occurs

```
Example with namespace A:  |a f1:1
Hidden Unit 0 = w_0_f1 * f1 + w_0_bias

When gradient is computed, it updates:
- w_0_f1 (the feature present)
- w_0_bias (shared across all examples)

Next example with namespace B: |b f2:1  
Hidden Unit 0 = w_0_f2 * f2 + w_0_bias

The w_0_bias was already influenced by the namespace A example,
creating implicit information transfer between namespaces.
```

### Potential Solutions (All Major Changes)

| Approach | Complexity | Description |
|----------|------------|-------------|
| Namespace-specific hidden layers | High | Create separate hidden layers per namespace |
| `--restricted_updates` flag | Medium | Only update weights for present features |
| Sparse gradient masking | Medium | Zero gradients for absent features during backprop |
| Namespace-aware architecture | High | Redesign network to respect namespace boundaries |

### Current Workarounds

1. **`--inpass`**: Adds direct input-to-output connections, reducing reliance on shared hidden layer
2. **Single namespace per example**: Ensure each example only contains features from one namespace
3. **Separate models**: Train separate models per namespace if isolation is critical

### Recommendation

This should be documented as a **known limitation** of `--nn` for namespace-sensitive use cases. A proper fix would require:
1. A new flag (e.g., `--nn_namespace_isolated`)
2. Significant changes to the forward/backward pass logic
3. Careful consideration of learning dynamics with sparse updates

The behavior is consistent with how standard fully-connected neural networks work - it's a design choice, not a bug.
